File: zapStatistic.py
#!/usr/bin/python
# coding: utf-8

import logging
import re, sys, argparse
from datetime import datetime
import csv


loglevel="debug"
#logging.basicConfig(filename='example.log',level=logging.DEBUG)
numeric_level = getattr(logging, loglevel.upper(), None)
print "numeric_level = " + str(numeric_level)
logging.basicConfig(level=numeric_level)

REG1 = "onGlobalKeyEvent.*CHANNEL_(UP|DOWN)" 
REG2 = "eXSERV_NOTIFICATION_EVENT_FIRST_FRAME"
resultTab = []

def parseArgs():
	logging.info('parse Args')
	parser = argparse.ArgumentParser(description='zap statsic')
	parser.add_argument('-f' , '--files', default=""    , help='file log to parse')
	parser.add_argument('-c', '--csvOutput'     , default="OFF"      , help='export result to csv file')

	## On parse et on update les variables globals du script
	args = parser.parse_args()
	globals().update(vars(args)) 
	return 0 

def getRegularDate(dateToCheck):
	ismatched = re.match("^([0-9]{4})([0-9]{2})([0-9]{2})T([0-9]{2})([0-9]{2})([0-9]{2})",dateToCheck)
	if  (ismatched):
		return [1, ismatched.group(1)+"/" + ismatched.group(2) +"/" + ismatched.group(3) + " " + ismatched.group(4) + ":" +ismatched.group(5) +":" + ismatched.group(6)]
        else:
    		return [0,"ERROR"];
    


def getZapDuration(start, end):
	date_format = "%H:%M:%S.%f"
	s = datetime.strptime(start, date_format)
	e = datetime.strptime(end, date_format)
	delta = e - s
	return delta.total_seconds() 

def parseLogFile():
	logging.info('parse log file')
	print "files are = "+ files+" and csv is = "+csvOutput 
        zapLineLog = ""
        parsingFileBreaked = 0
	newZapCount = 0	

	for fileName in files.split("#"):
		file = open(fileName, "r") 
		print fileName + " : is currently parsing"; 	
		parsingFileBreaked = 0;
		currentStatisticResult = {}
		newZapCount = 0

		for line in file: 
			if re.findall(REG1, line):
				zapLineLog =  line
			
			if(len(zapLineLog) and re.findall(REG2, line)):
			      liveDisplayLine = line.split(" ");
      			      startZapLine = zapLineLog.split(" ");
                             
		              isRegularStartZapTime = (getRegularDate(startZapLine[0]))[0]
              		      isRegularLiveDispaly = (getRegularDate(liveDisplayLine[0]))[0]

		              # break file parsing if it's wrongly timestamped  
              		      if ((isRegularStartZapTime+isRegularLiveDispaly) != 2) :
		                   parsingFileBreaked = 1;
                		   print "ERROR : File is wrongly timestamped\n" 
		                   break
			      tmpDic = {}
			      tmpDic['startZapDate'] = (getRegularDate(startZapLine[0]))[1]   
			      tmpDic['zappingDuration'] = getZapDuration(startZapLine[2] , liveDisplayLine[2]);
		              currentStatisticResult[newZapCount] = tmpDic
			      # increment the counter
		              newZapCount += 1
      
		              # reset the zapLineLog
		      	      zapLineLog = ""

              
		file.close()
		if parsingFileBreaked == 0:
         		 currentStatisticResult['totalZap'] = newZapCount ; 
		         # save the current statistic array 
		         resultTab.append(currentStatisticResult);



def displayZapStatistic():
	logging.info('display Result')
	size = len(resultTab) 
	filesName = files.split("#")
	cpt = 0	
	
	while cpt < size :
		if(csvOutput == "ON"):
      		 	tmpName = (filesName[cpt])+".csv"
			f = open(tmpName, "wb")
			try:
				c = csv.writer(f)
				c.writerow( ('Date', 'Duration') )
			finally:
				logging.info('can open the csv file')	


		hashSize = (resultTab[cpt])['totalZap']
		jj = 0
		while jj < hashSize:
        		if(csvOutput == "OFF"):
              			print  str(jj)+"   "+(resultTab[cpt])[jj]['startZapDate']+"   "+str((resultTab[cpt])[jj]['zappingDuration'])+ " sec";
	           	else:     
 				c.writerow( ((resultTab[cpt])[jj]['startZapDate'] , str((resultTab[cpt])[jj]['zappingDuration']) ) ) 
			jj += 1
		
		if csvOutput == "ON" :
			f.close()
		cpt +=1          
       


def main(argv):
	parseArgs()
	parseLogFile()
	displayZapStatistic()


main(sys.argv[1:])

File: zapStatistic.pl
#!/usr/bin/perl 

use strict;
use warnings;
use DateTime;
use Getopt::Long;

#die "usage: /usr/bin/perl $0 file1 file2.\n" if(($#ARGV + 1) == 0) ; 

just for test
#  this table will stock all statistic hash of log files
our @GlobalStatisticZapResult = ();

# ARGV passed by the script
my $csvExport = "OFF"; # by default the script does't generate the csv files
my @filesNameArr = () ;



parseParams() || die "uage: /usr/bin/perl $0 [option] -csv ON/OFF -F file1*file2*file...\nExp : perl $0 -F file1.log#file2.log \n      Perl $0 -csv ON -F file1.log#file2.log";
# parse log files
parseLogFiles();
displayZapStatistic();



sub parseParams{
   my $files = "";
   GetOptions(
    "csv=s"   => \$csvExport,   
    "F=s"     => \$files,
   );
   #print "files = ".$files."\n";
   if(length($files) > 0){
     @filesNameArr = split /#/, $files;   
   }
   return @filesNameArr; 
}

sub parseLogFiles{
   
 
  foreach my $file (@filesNameArr) {
       $newZapCount = 0;	
       $parsingFileBreaked= 0;
       # new instanciation for each file parsing 
       my %currentSatisticZapResult;
       
       print $file." is currently parsing\n"; 	
       open my $fh, '<:encoding(UTF-8)', $file or die;
       while (my $line = <$fh>) {
          
          if ($line =~ /$REG1/){ 
             $zapLineLog = $line; 
           }
      		
          if(length($zapLineLog) != 0 && $line =~ /$REG2/) {
      	   # retrieve the date of live dispaly
              $currentLine = $line	;
      	      my @liveDisplayLine = split / /, $currentLine;
      	      my @startZapLine = split / /, $zapLineLog;
                             
              my $isRegularStartZapTime = (getRegularDate($startZapLine[0]))[0];
              my $isRegularLiveDispaly = (getRegularDate($liveDisplayLine[0]))[0];  
              # break file parsing if it's wrongly timestamped  
              if (($isRegularStartZapTime+$isRegularLiveDispaly) != 2){
                   $parsingFileBreaked = 1;
                   print "ERROR : File is wrongly timestamped\n" ; 
                   last;
              }
              $currentSatisticZapResult{$newZapCount}{'startZapDate'} =  (getRegularDate($startZapLine[0]))[1] ;
              $currentSatisticZapResult{$newZapCount}{'zappingDuration'} = getZapDuration($startZapLine[2] , $liveDisplayLine[2]);
              # increment the counter
              $newZapCount++;
      
              # reset the zapLineLog
      	      $zapLineLog = "";
          }
       }#end wile
       close($file);
       if($parsingFileBreaked == 0){
          # save the total zap in the parsed log file
          $currentSatisticZapResult{'info'}{'totalZap'} = $newZapCount ; 
          # save the current statistic array 
          push(@GlobalStatisticZapResult, \%currentSatisticZapResult);
       }
  }#end foreach
} 


sub displayZapStatistic{
    my $size = @GlobalStatisticZapResult ;
    for(my $cpt = 0 ; $cpt < $size ; $cpt++){
       if($csvExport eq "ON"){
  	    openCSVfiles($filesNameArr[$cpt]);
       }
       
       my $hashSize = $GlobalStatisticZapResult[$cpt]->{'info'}->{'totalZap'};
       for(my $jj = 0 ; $jj < $hashSize ; $jj++) {
           if($csvExport eq "OFF"){  
              print  $jj.'   '.$GlobalStatisticZapResult[$cpt]->{$jj}->{'startZapDate'}.'   '.$GlobalStatisticZapResult[$cpt]->{$jj}{'zappingDuration'}." sec\n";
           }else{
              print FH $GlobalStatisticZapResult[$cpt]->{$jj}->{'startZapDate'}.','.$GlobalStatisticZapResult[$cpt]->{$jj}{'zappingDuration'}."\n";
           }
       }
       
       closeCSVfiles()  if ($csvExport eq "ON");
       print "/////////   END OF PARSING the Log File : $filesNameArr[$cpt]   //////////\n";
   }
}


sub getRegularDate {
    my $stringToConvert = shift; 
    if  ($stringToConvert =~ /^([0-9]{4})([0-9]{2})([0-9]{2})T([0-9]{2})([0-9]{2})([0-9]{2})/) {	
	 return (1,$1.'/'.$2.'/'.$3.' '.$4.':'.$5.':'.$6);
    }
    else 
    {
        return (0,"ERROR");
    }
}

sub openCSVfiles{
    my $parsedLogName = shift;
    open (FH, ">$parsedLogName.csv") or die "$!";
    # add first line 
    print FH "Date, Duration\n";
}

sub closeCSVfiles{
   close(FH);
}

sub getZapDuration{
    my ($startTime , $endTime) = @_; #"05:07:53.249"
    my $diff =  to_milli($endTime) - to_milli($startTime);
    return ($diff/1000);
}

sub to_milli {
    my @components = split /[:\.]/, shift;
    return (($components[0] * 60 + $components[1]) * 60 + $components[+2]) * 1000 + $components[3];
}


